compiler/frontend/src/mir_to_lir.rs

```diff
@@ -0,0 +1,391 @@
+use crate::{
+    error::CompilerError,
+    hir,
+    id::CountableId,
+    lir::{self, Lir},
+    mir::{self},
+    mir_optimize::OptimizeMir,
+    module::Module,
+    string_to_rcst::ModuleError,
+    TracingConfig,
+};
+use itertools::Itertools;
+use rustc_hash::{FxHashMap, FxHashSet};
+use std::{mem, sync::Arc};
+
+#[salsa::query_group(MirToLirStorage)]
+pub trait MirToLir: OptimizeMir {
+    fn lir(&self, module: Module, tracing: TracingConfig) -> LirResult;
+}
+
+pub type LirResult = Result<(Arc<Lir>, Arc<FxHashSet<CompilerError>>), ModuleError>;
+
+fn lir(db: &dyn MirToLir, module: Module, tracing: TracingConfig) -> LirResult {
+    let (mir, _, errors) = db.optimized_mir(module.clone(), tracing)?;
+
+    let mut context = LoweringContext::default();
+    context.compile_function(
+        FxHashSet::from_iter([hir::Id::new(module, vec![])]),
+        &[],
+        &[],
+        mir::Id::from_usize(0),
+        &mir.body,
+    );
+    let lir = Lir::new(context.constants, context.bodies);
+
+    Ok((Arc::new(lir), errors))
+}
+
+#[derive(Clone, Debug, Default)]
+struct LoweringContext {
+    constants: lir::Constants,
+    constant_mapping: FxHashMap<mir::Id, lir::ConstantId>,
+    bodies: lir::Bodies,
+    current_body: CurrentBody,
+    last_constant: Option<mir::Id>,
```

Comment:
I think we can move some of this around so that `mem::replace` is no longer necessary. What if the `LoweringContext` contained only things where data is never removed (`constants`, `bodies`, `constant_mapping`) and the `last_constant` is moved into the `CurrentBody` and set during `finish`?